lib/spack/spack/environment/environment.py

```diff
@@ -1034,6 +1177,50 @@ def scope_name(self):
         """Name of the config scope of this environment's manifest file."""
         return self.manifest.scope_name
 
+    def include_concrete_envs(self):
+        """Copy and save the included envs' specs internally"""
+
+        lockfile_meta = None
+        root_hash_seen = set()
+        concrete_hash_seen = set()
+        self.included_concrete_specs = {}
+
+        for env_path in self.include_concrete:
+            # Check that environment exists
+            if not is_env_dir(env_path):
+                raise SpackEnvironmentError(f"Unable to find env at {env_path}")
+
+            env = Environment(env_path)
+
+            with open(env.lock_path) as f:
+                lockfile_as_dict = env._read_lockfile(f)
+
+            # Lockfile_meta must match each env and use at least format version 5
+            if lockfile_meta is None:
+                lockfile_meta = lockfile_as_dict["_meta"]
+            elif lockfile_meta != lockfile_as_dict["_meta"]:
+                raise SpackEnvironmentError("All lockfile _meta values must match")
+            elif lockfile_meta["lockfile-version"] < 5:
+                raise SpackEnvironmentError("The lockfile format must be at version 5 or higher")
+
+            # Copy unique root specs from env
+            self.included_concrete_specs[env_path] = {"roots": []}
```

Comment:
It looks like this is only iterating over the "local" specs in the included environment, won't this cause a problem if we have a chain of includes?

Reply:
I thought the property name was confusing so renamed it to `included_concrete_spec_data` to reflect what it contains.

I also finished implementing  `set_included_concretized_user_specs` to include `add_specs` to populate the `included_concretized_order` and `included_concretized_user_specs` properties with included/nested concrete includes.